tests/src/test/scala/com/nvidia/spark/rapids/spill/SpillFrameworkSuite.scala

```diff
@@ -0,0 +1,1103 @@
+            TestUtils.compareBatches(decompressedExpected, decompressed)
+          }
+        }
+      }
+    }
+  }
+
+  test("a compressed batch can be added and recovered after being spilled to disk") {
+    val ct = buildCompressedBatch(0, 1000)
+    withResource(decompressBach(ct)) { decompressedExpected =>
+      withResource(SpillableCompressedColumnarBatchHandle(ct)) { handle =>
+        assert(handle.spillable)
+        SpillFramework.stores.deviceStore.spill(handle.approxSizeInBytes)
+        assert(!handle.spillable)
+        SpillFramework.stores.hostStore.spill(handle.approxSizeInBytes)
+        assert(handle.dev.isEmpty)
+        assert(handle.host.isDefined)
+        assert(handle.host.get.host.isEmpty)
+        assert(handle.host.get.disk.isDefined)
+        withResource(handle.materialize()) { materialized =>
+          withResource(decompressBach(materialized)) { decompressed =>
+            TestUtils.compareBatches(decompressedExpected, decompressed)
+          }
+        }
+      }
+    }
+  }
+
+
+  test("a second handle prevents buffer to be spilled") {
+    val buffer = DeviceMemoryBuffer.allocate(123)
+    val handle1 = SpillableDeviceBufferHandle(buffer)
+    // materialize will incRefCount `buffer`. This looks a little weird
+    // but it simulates aliasing as it happens in real code
+    val handle2 = SpillableDeviceBufferHandle(handle1.materialize())
+
+    withResource(handle1) { _ =>
+      withResource(handle2) { _ =>
+        assertResult(2)(handle1.dev.get.getRefCount)
+        assertResult(2)(handle2.dev.get.getRefCount)
+        assertResult(false)(handle1.spillable)
+        assertResult(false)(handle2.spillable)
+      }
+      assertResult(1)(handle1.dev.get.getRefCount)
+      assertResult(true)(handle1.spillable)
+    }
+  }
+
+  test("removing handle releases buffer resources in all stores") {
+    val handle = SpillableDeviceBufferHandle(DeviceMemoryBuffer.allocate(123))
+    withResource(handle) { _ =>
+      assertResult(1)(SpillFramework.stores.deviceStore.numHandles)
+      assertResult(0)(SpillFramework.stores.hostStore.numHandles)
+      assertResult(0)(SpillFramework.stores.diskStore.numHandles)
+
+      assertResult(123)(SpillFramework.stores.deviceStore.spill(123)) // spill to host memory
+      assertResult(0)(SpillFramework.stores.deviceStore.numHandles)
+      assertResult(1)(SpillFramework.stores.hostStore.numHandles)
+      assertResult(0)(SpillFramework.stores.diskStore.numHandles)
+      assert(handle.dev.isEmpty)
+      assert(handle.host.isDefined)
+      assert(handle.host.get.host.isDefined)
+
+      assertResult(123)(SpillFramework.stores.hostStore.spill(123)) // spill to disk
+      assertResult(0)(SpillFramework.stores.deviceStore.numHandles)
+      assertResult(0)(SpillFramework.stores.hostStore.numHandles)
+      assertResult(1)(SpillFramework.stores.diskStore.numHandles)
+      assert(handle.dev.isEmpty)
+      assert(handle.host.isDefined)
```

Comment:
can you explain/add a comment as to why handle.host is defined but handle.host.get.host is not? I'm sure there's a good reason but it's not obvious

Reply:
The way we have architected the spill handles is that they can cascade device->host->disk. But, the device handles don't have a host AND disk handles, they just have a host handle. If the host handle itself spilled, or could not fit on host to begin with, its host component is empty, and its disk component is set. So the view of the world from the handle's perspective is either set in the place where they are supposed to be, or they have a handle to something that can help them find the object later.

Reply:
Ah I see, it's kind of like how the spill stores had a `next` store reference